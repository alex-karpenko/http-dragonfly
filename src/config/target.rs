use hyper::Uri;
use jaq_interpret::{Ctx, Filter, FilterT, ParseCtx, RcIter, Val};
use serde::{
    de::{self, Visitor},
    Deserialize, Deserializer,
};
use std::time::Duration;
use tracing::error;

use crate::errors::HttpDragonflyError;

use super::{headers::HeaderTransform, response::ResponseStatus, ConfigValidator};

const DEFAULT_TARGET_TIMEOUT_SEC: u64 = 60;

pub type TargetConfigList = Vec<TargetConfig>;

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct TargetConfig {
    id: Option<String>,
    url: String,
    headers: Option<Vec<HeaderTransform>>,
    body: Option<String>,
    #[serde(
        with = "humantime_serde",
        default = "TargetConfig::default_target_timeout"
    )]
    timeout: Duration,
    #[serde(default)]
    on_error: TargetOnErrorAction,
    error_status: Option<ResponseStatus>,
    condition: Option<TargetConditionConfig>,
}

impl TargetConfig {
    fn default_target_timeout() -> Duration {
        Duration::from_secs(DEFAULT_TARGET_TIMEOUT_SEC)
    }

    pub fn id(&self) -> String {
        if let Some(id) = &self.id {
            id.clone()
        } else {
            format!("TARGET-{}", self.url)
        }
    }

    pub fn uri(&self) -> Result<Uri, HttpDragonflyError> {
        self.url
            .parse()
            .map_err(|e| HttpDragonflyError::InvalidConfig {
                cause: format!("invalid url `{}`: {e}", self.url),
            })
    }

    pub fn host(&self) -> String {
        if let Ok(uri) = self.uri() {
            uri.host().unwrap_or("").to_lowercase()
        } else {
            String::new()
        }
    }

    pub fn url(&self) -> &str {
        self.url.as_ref()
    }

    pub fn headers(&self) -> &Option<Vec<HeaderTransform>> {
        &self.headers
    }

    pub fn body(&self) -> &Option<String> {
        &self.body
    }

    pub fn timeout(&self) -> Duration {
        self.timeout
    }

    pub fn on_error(&self) -> &TargetOnErrorAction {
        &self.on_error
    }

    pub fn error_status(&self) -> Option<u16> {
        self.error_status
    }

    pub fn condition(&self) -> &Option<TargetConditionConfig> {
        &self.condition
    }
}

#[derive(Deserialize, Debug, Default)]
#[serde(deny_unknown_fields, rename_all = "snake_case")]
pub enum TargetOnErrorAction {
    #[default]
    Propagate,
    Status,
    Drop,
}

#[derive(Debug)]
pub enum TargetConditionConfig {
    Default,
    Filter(ConditionFilter),
}

impl TargetConditionConfig {
    pub fn from_str(value: &str) -> Result<Self, HttpDragonflyError> {
        match value {
            "default" => Ok(TargetConditionConfig::Default),
            _ => Ok(TargetConditionConfig::Filter(ConditionFilter::from_str(
                value,
            )?)),
        }
    }
}

impl From<&str> for TargetConditionConfig {
    fn from(value: &str) -> Self {
        Self::from_str(value).expect("unable to parse conditional expression")
    }
}

impl<'de> Deserialize<'de> for TargetConditionConfig {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct TargetConditionConfigVisitor;
        impl<'de> Visitor<'de> for TargetConditionConfigVisitor {
            type Value = TargetConditionConfig;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str(
                    "conditional expression in JQ-like style that returns false/true value",
                )
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                TargetConditionConfig::from_str(v).map_err(|e| E::custom(e))
            }
        }

        deserializer.deserialize_string(TargetConditionConfigVisitor)
    }
}

#[derive(Debug)]
pub struct ConditionFilter {
    filter: Filter,
}

impl From<&str> for ConditionFilter {
    fn from(value: &str) -> Self {
        Self::from_str(value).expect("unable to parse conditional expression")
    }
}
impl ConditionFilter {
    pub fn run(&self, input: serde_json::value::Value) -> bool {
        let inputs = RcIter::new(core::iter::empty());
        let out = self.filter.run((Ctx::new([], &inputs), Val::from(input)));

        let out: Vec<String> = out
            .map(|v| format!("{}", v.unwrap_or(Val::Bool(false))))
            .collect();

        out.len() == 1 && out[0] == "true"
    }

    pub fn from_str(value: &str) -> Result<Self, HttpDragonflyError> {
        let mut defs = ParseCtx::new(Vec::new());
        let (f, errs) = jaq_parse::parse(value, jaq_parse::main());
        if !errs.is_empty() {
            errs.iter()
                .for_each(|e| error!("unable to parse conditional expression: {e}"));
            return Err(HttpDragonflyError::InvalidConfig {
                cause: errs[0].to_string(),
            });
        }
        if let Some(f) = f {
            let filter = defs.compile(f);
            Ok(ConditionFilter { filter })
        } else {
            Err(HttpDragonflyError::InvalidConfig {
                cause: "invalid conditional expression".into(),
            })
        }
    }
}

impl ConfigValidator for TargetConfig {
    fn validate(&self) -> Result<(), HttpDragonflyError> {
        // Validate URIs
        self.uri()?;

        // Validate target's error response override
        match self.on_error() {
            TargetOnErrorAction::Propagate | TargetOnErrorAction::Drop => {
                if self.error_status().is_some() {
                    return Err(HttpDragonflyError::InvalidConfig {
                        cause: format!(
                            "`error_status` can be set if `on_error` is `status` only, target `{}`",
                            self.id()
                        ),
                    });
                }
            }
            TargetOnErrorAction::Status => {
                if self.error_status().is_none() {
                    return Err(HttpDragonflyError::InvalidConfig {
                        cause: format!(
                            "`error_status` should be set if `on_error` is `status`, target `{}`",
                            self.id()
                        ),
                    });
                }
            }
        }

        Ok(())
    }
}

impl ConfigValidator for [TargetConfig] {
    fn validate(&self) -> Result<(), HttpDragonflyError> {
        // Validate each target
        for target in self {
            target.validate()?;
        }

        // Make sure all targets have unique ID
        let unique_targets_count = self.iter().map(TargetConfig::id).count();
        if unique_targets_count != self.len() {
            return Err(HttpDragonflyError::InvalidConfig {
                cause: "all target IDs of the listener should be unique".into(),
            });
        }

        Ok(())
    }
}
