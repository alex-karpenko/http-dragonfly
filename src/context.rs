use hyper::{http::request::Parts, Body, Response};
use once_cell::sync::OnceCell;
use regex::Regex;
use std::collections::HashMap;
use std::env;
use std::net::SocketAddr;

use crate::config::target::TargetConfig;

const CTX_APP_NAME: &str = env!("CARGO_PKG_NAME");
const CTX_APP_VERSION: &str = env!("CARGO_PKG_VERSION");

pub type ContextMap = HashMap<String, String>;

#[derive(Debug)]
pub struct Context<'a> {
    own: ContextMap,
    parent: Option<&'a Context<'a>>,
}

impl<'a> Context<'a> {
    pub fn root(root_env: impl RootEnvironment) -> &'static Context<'a> {
        static ROOT_CONTEXT: OnceCell<Context> = OnceCell::new();

        let ctx = ROOT_CONTEXT.get_or_init(|| {
            let mut ctx = ContextMap::new();

            ctx.insert("CTX_APP_NAME".into(), CTX_APP_NAME.into());
            ctx.insert("CTX_APP_VERSION".into(), CTX_APP_VERSION.into());
            ctx.extend(root_env.get_environment());

            Context {
                own: ctx,
                parent: None,
            }
        });

        ctx
    }

    pub fn with(&self, own: ContextMap) -> Context {
        Context {
            own,
            parent: Some(self),
        }
    }

    pub fn get(&self, var: &String) -> Option<&String> {
        // Try own context
        if let Some(value) = self.own.get(var) {
            Some(value)
            // or get parent if it's
        } else if let Some(parent) = &self.parent {
            // and try parent
            if let Some(value) = parent.get(var) {
                Some(value)
            } else {
                None
            }
        } else {
            None
        }
    }

    pub fn iter(&'a self) -> Box<dyn Iterator<Item = (&'a String, &'a String)> + 'a> {
        let iter = Box::new(self.own.iter());
        Box::new(Iter {
            ctx: self,
            iter,
            finished: false,
        })
    }

    pub fn with_request(
        &'a self,
        addr: &'a SocketAddr,
        req: &'a Parts,
        listener_name: String,
    ) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_LISTENER_NAME
        // CTX_REQUEST_SOURCE_IP
        // CTX_REQUEST_METHOD
        // CTX_REQUEST_HOST
        // CTX_REQUEST_PATH
        // CTX_REQUEST_QUERY
        own.insert("CTX_LISTENER_NAME".into(), listener_name);
        own.insert("CTX_REQUEST_SOURCE_IP".into(), addr.ip().to_string());
        own.insert("CTX_REQUEST_METHOD".into(), req.method.to_string());
        own.insert("CTX_REQUEST_PATH".into(), req.uri.path().to_string());
        if let Some(host) = req.uri.host() {
            own.insert("CTX_REQUEST_HOST".into(), host.to_lowercase());
        }
        if let Some(query) = req.uri.query() {
            own.insert("CTX_REQUEST_QUERY".into(), query.to_lowercase());
        }

        // CTX_REQUEST_HEADERS_<UPPERCASE_HEADER_NAME>
        req.headers.iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_REQUEST_HEADERS_{n}"), v);
        });

        self.with(own)
    }

    pub fn with_target(&'a self, cfg: &'a TargetConfig) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_TARGET_ID
        // CTX_TARGET_HOST
        own.insert("CTX_TARGET_ID".into(), cfg.id());
        own.insert("CTX_TARGET_HOST".into(), cfg.host());

        self.with(own)
    }

    pub fn with_response(&'a self, resp: &Response<Body>) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_RESPONSE_HEADERS_<UPPERCASE_HEADER_NAME>
        // CTX_RESPONSE_STATUS
        own.insert("CTX_RESPONSE_STATUS".into(), resp.status().to_string());
        resp.headers().iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_RESPONSE_HEADERS_{n}"), v);
        });

        self.with(own)
    }
}

pub struct Iter<'a> {
    ctx: &'a Context<'a>,
    iter: Box<dyn Iterator<Item = (&'a String, &'a String)> + 'a>,
    finished: bool,
}

impl<'a> Iterator for Iter<'a> {
    type Item = (&'a String, &'a String);

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(next) = self.iter.next() {
            Some(next)
        } else if self.ctx.parent.is_none() {
            None
        } else if !self.finished {
            self.finished = true;
            self.iter = self.ctx.parent.unwrap().iter();
            self.iter.next()
        } else {
            None
        }
    }
}

pub trait RootEnvironment {
    fn get_environment(&self) -> ContextMap;
}

pub struct RootOsEnvironment<'a> {
    env_mask_regex: &'a str,
}

impl<'a> RootEnvironment for RootOsEnvironment<'a> {
    fn get_environment(&self) -> ContextMap {
        let mut env_ctx = ContextMap::new();
        let re = Regex::new(self.env_mask_regex).expect("invalid environment filter regex");
        env::vars().for_each(|(k, v)| {
            if re.is_match(&k) {
                env_ctx.insert(k, v);
            }
        });

        env_ctx
    }
}

impl<'a> RootOsEnvironment<'a> {
    pub fn new(env_mask_regex: &'a str) -> Self {
        Self { env_mask_regex }
    }
}
