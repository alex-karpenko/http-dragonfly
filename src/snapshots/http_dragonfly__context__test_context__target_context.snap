---
source: src/context.rs
expression: target_ctx
---
Context(
  own: {
    "CTX_TARGET_HOST": "www.google.com",
    "CTX_TARGET_ID": "TEST-TARGET-ID",
  },
  parent: Some(Context(
    own: {
      "CTX_APP_NAME": "http-dragonfly",
      "CTX_APP_VERSION": "0.3.6",
      "TEST_ENV_HEADER_TO_ADD": "TEST_ENV_HEADER_VALUE",
      "TEST_ENV_KEY": "TEST_ENV_VALUE",
    },
    parent: None,
  )),
)
